File: minimum_failing_path.java
class Solution {
    public int minFallingPathSum(int[][] matrix) {
        int[][] dp=new int[matrix.length][matrix.length];
        for(int i=0; i<matrix.length; i++)
        {
            dp[0][i]=matrix[0][i];
        }
        for(int i=1; i<matrix.length; i++)
        {
            for(int j=0; j<matrix[0].length; j++)
            {
                if(j>0 && j<matrix.length-1)
                {
                    dp[i][j] = matrix[i][j] + Math.min(Math.min(dp[i-1][j],dp[i-1][j-1]),dp[i-1][j+1]);
                }
                else if(j==0)
                {
                    dp[i][j] = matrix[i][j] + Math.min(dp[i-1][j],dp[i-1][j+1]);
                }
                else if(j==matrix.length-1)
                {
                    dp[i][j] = matrix[i][j] + Math.min(dp[i-1][j],dp[i-1][j-1]);
                }
            }
        }
        int sol=100*matrix.length*matrix.length;
        for(int k=0; k<matrix.length; k++)
        {
            sol=Math.min(sol,dp[matrix.length-1][k]);
        }
        return sol;
    }
}

File: minimum_failing_path.py
"""
Time Complexity : O(m*n) where m is the number of rows and n is the number of cols of matrix
Space Complexity : O(1) as no auxillary data structure is used
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No
"""


class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        if len(matrix) == 0:
            return 0

        m = len(matrix)
        n = len(matrix[0])
        minSum = float('inf')
        for i in range(1, m):
            for j in range(0, n):
                if j == 0:
                    matrix[i][j] = min(
                        matrix[i-1][j], matrix[i-1][j+1]) + matrix[i][j]
                elif j == n-1:
                    matrix[i][j] = min(
                        matrix[i-1][j], matrix[i-1][j-1]) + matrix[i][j]
                else:
                    matrix[i][j] = min(
                        matrix[i-1][j], matrix[i-1][j-1], matrix[i-1][j+1]) + matrix[i][j]

        for j in range(n):
            minSum = min(minSum, matrix[m-1][j])

        return minSum
